File: app/api/src/Astra.Api/Llm/Archetypes/dotnet8/canonical-delphi-idsmtp/src/IdSMTPMin.cs
// SPDX-Spec: delphi/TIdSMTPMin (signed)
// SPDX-Archetype: canonical-delphi-idsmtp
//
// .NET 8 projection of Delphi's TIdSMTPMin — Indy's minimal SMTP client. Each
// public surface carries a [SpecClaim(...)] attribute citing the signed claim id;
// reviewers can map the C# back to the spec without leaving the IDE. The
// implementation BODY is intentionally TODO — the scaffold is a contract, not a
// translation. The signed spec is the source of truth; the implementer fills the
// body and the equivalence sidecar (Phase 9.0.f / fpc-sidecar) plus the property-
// test sidecar (9.0.g) confirm behavioural parity.

using System.Net.Sockets;

namespace Demo.IdSMTP;

[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
public sealed class SpecClaimAttribute(string id) : Attribute
{
    public string Id { get; } = id;
}

/// <summary>
/// .NET 8 projection of Delphi's <c>TIdSMTPMin</c>. Owns a TCP connection to an
/// SMTP server and exposes the minimum command set needed for Phase 9.0's demo
/// (Connect → HELO → MAIL FROM → RCPT TO → DATA → QUIT).
/// </summary>
/// <remarks>
/// Object-lifetime model (OL-1): this class is <b>Owned</b> by its caller. The
/// caller must Dispose() it; the underlying TcpClient is closed in Dispose().
/// </remarks>
[SpecClaim("II-1")]
[SpecClaim("OL-1")]
public sealed class IdSMTPMin : IIdComponent
{
    private readonly TcpClient _socket = new();
    private string _host = string.Empty;
    private int _port = 25;
    private bool _disposed;

    /// <summary>Remote host. PA-1: property accessor, no side effects on read.</summary>
    [SpecClaim("PA-1")]
    public string Host
    {
        get => _host;
        set { ThrowIfDisposed(); _host = value ?? string.Empty; }
    }

    /// <summary>Remote port. INV-1: 1 ≤ Port ≤ 65535 enforced on set.</summary>
    [SpecClaim("INV-1")]
    [SpecClaim("PA-2")]
    public int Port
    {
        get => _port;
        set
        {
            ThrowIfDisposed();
            if (value is < 1 or > 65535)
                throw new ArgumentOutOfRangeException(nameof(value), "Port must be in [1, 65535].");
            _port = value;
        }
    }

    /// <inheritdoc/>
    public bool Connected => !_disposed && _socket.Connected;

    /// <summary>Fired on connection-state transitions. EH-1.</summary>
    [SpecClaim("EH-1")]
    public event EventHandler<IdStatusEventArgs>? Status;

    /// <summary>Fired once the socket is fully closed. EH-2.</summary>
    [SpecClaim("EH-2")]
    public event EventHandler? Disconnect;

    /// <inheritdoc/>
    [SpecClaim("INV-1")]
    [SpecClaim("SE-1")]
    public void Open()
    {
        ThrowIfDisposed();
        if (Connected) return; // INV-1: idempotent on already-open.
        // TODO(impl): _socket.Connect(_host, _port);
        // TODO(impl): RaiseStatus(IdConnectionStatus.Connected, $"{_host}:{_port}");
        // TODO(impl): Read greeting line (220 ...) — emit EC-1 if not 220.
        throw new NotImplementedException("Connect body — fill from signed spec INV-1, SE-1.");
    }

    /// <inheritdoc/>
    [SpecClaim("INV-2")]
    public void Close()
    {
        if (_disposed || !_socket.Connected) return; // INV-2: idempotent.
        // TODO(impl): send "QUIT\r\n" if a command is in flight (EC-2).
        // TODO(impl): _socket.Close();
        // TODO(impl): RaiseStatus(IdConnectionStatus.Disconnected, "QUIT");
        Disconnect?.Invoke(this, EventArgs.Empty);
    }

    /// <summary>Send a raw SMTP command. SE-1: writes to the socket.</summary>
    [SpecClaim("SE-1")]
    [SpecClaim("EC-1")]
    public string SendCmd(string command, int expectedCode)
    {
        ThrowIfDisposed();
        if (!Connected) throw new InvalidOperationException("Not connected. Call Open() first (INV-1).");
        // TODO(impl): write `command + "\r\n"`; read response line; parse 3-digit code.
        // TODO(impl): if response.code != expectedCode → throw IdSmtpReplyException (EC-1).
        throw new NotImplementedException("SendCmd body — fill from signed spec SE-1, EC-1.");
    }

    public void Dispose()
    {
        if (_disposed) return;
        try { Close(); } finally { _socket.Dispose(); _disposed = true; }
    }

    private void ThrowIfDisposed()
    {
        if (_disposed) throw new ObjectDisposedException(nameof(IdSMTPMin));
    }
}

File: app/api/src/Astra.Api/Llm/Archetypes/java-spring/canonical-delphi-idsmtp/src/main/java/com/example/idsmtp/IdSMTPMin.java
// SPDX-Spec: delphi/TIdSMTPMin (signed)
// SPDX-Archetype: canonical-delphi-idsmtp
//
// Java 17 + Spring Boot projection of Delphi's TIdSMTPMin. Every public surface
// carries a @SpecClaim annotation citing the signed claim id; reviewers can map
// the Java back to the spec without leaving the IDE. The implementation BODY is
// intentionally TODO — the scaffold is a contract, not a translation. The signed
// spec is the source of truth; the implementer fills the body and the equivalence
// sidecar (Phase 9.0.f / fpc-sidecar) plus the property-test sidecar (9.0.g)
// confirm behavioural parity.

package com.example.idsmtp;

import java.io.IOException;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import org.springframework.stereotype.Component;

@Retention(RetentionPolicy.RUNTIME)
@Repeatable(SpecClaim.Container.class)
@interface SpecClaim {
    String value();
    @Retention(RetentionPolicy.RUNTIME)
    @interface Container { SpecClaim[] value(); }
}

/**
 * Java 17 projection of Delphi's <code>TIdSMTPMin</code>. Owns a TCP connection
 * to an SMTP server and exposes the minimum command set needed for Phase 9.0's
 * demo (Connect → HELO → MAIL FROM → RCPT TO → DATA → QUIT).
 *
 * <p>Object-lifetime model (OL-1): this class is <b>Owned</b> by its caller.
 * The caller must call {@link #close()} (or use try-with-resources); the
 * underlying socket is closed in close().
 */
@Component
@SpecClaim("II-1")
@SpecClaim("OL-1")
public class IdSMTPMin implements IdComponent {

    private Socket socket;
    private String host = "";
    private int port = 25;
    private boolean closed;
    private final List<IdStatusListener> listeners = new ArrayList<>();

    // ---- getter/setter pair (Delphi property → Java accessors) -----------------

    /** Remote host. PA-1: property accessor, no side effects on read. */
    @SpecClaim("PA-1")
    public String getHost() { return host; }

    @SpecClaim("PA-1")
    public void setHost(String host) {
        throwIfClosed();
        this.host = Objects.requireNonNullElse(host, "");
    }

    /** Remote port. INV-1: 1 ≤ port ≤ 65535 enforced on set. */
    @SpecClaim("INV-1")
    @SpecClaim("PA-2")
    public int getPort() { return port; }

    @SpecClaim("INV-1")
    @SpecClaim("PA-2")
    public void setPort(int port) {
        throwIfClosed();
        if (port < 1 || port > 65535) {
            throw new IllegalArgumentException("Port must be in [1, 65535].");
        }
        this.port = port;
    }

    // ---- listener registration (Delphi event → Java listener) -------------------

    /** Subscribe to connection-state transitions. EH-1. */
    @SpecClaim("EH-1")
    public void addStatusListener(IdStatusListener l) { listeners.add(l); }

    public void removeStatusListener(IdStatusListener l) { listeners.remove(l); }

    private void fireStatus(IdConnectionStatus s, String detail) {
        var ev = new IdStatusEvent(s, detail);
        for (var l : listeners) l.onStatus(ev);
    }

    // ---- IdComponent contract ---------------------------------------------------

    @Override
    public boolean isConnected() {
        return !closed && socket != null && socket.isConnected();
    }

    @Override
    @SpecClaim("INV-1")
    @SpecClaim("SE-1")
    public void open() {
        throwIfClosed();
        if (isConnected()) return; // INV-1: idempotent on already-open.
        // TODO(impl): socket = new Socket(host, port);
        // TODO(impl): fireStatus(IdConnectionStatus.CONNECTED, host + ":" + port);
        // TODO(impl): read greeting line (220 ...) — emit EC-1 if not 220.
        throw new UnsupportedOperationException("Connect body — fill from signed spec INV-1, SE-1.");
    }

    @Override
    @SpecClaim("INV-2")
    public void close() {
        if (closed || socket == null) return; // INV-2: idempotent.
        try {
            // TODO(impl): send "QUIT\r\n" if a command is in flight (EC-2).
            socket.close();
            fireStatus(IdConnectionStatus.DISCONNECTED, "QUIT");
        } catch (IOException e) {
            throw new RuntimeException("Failed to close socket", e);
        } finally {
            closed = true;
        }
    }

    /** Send a raw SMTP command. SE-1: writes to the socket. */
    @SpecClaim("SE-1")
    @SpecClaim("EC-1")
    public String sendCmd(String command, int expectedCode) {
        throwIfClosed();
        if (!isConnected()) {
            throw new IllegalStateException("Not connected. Call open() first (INV-1).");
        }
        // TODO(impl): write command + "\r\n"; read response line; parse 3-digit code.
        // TODO(impl): if response.code != expectedCode → throw IdSmtpReplyException (EC-1).
        throw new UnsupportedOperationException("sendCmd body — fill from signed spec SE-1, EC-1.");
    }

    private void throwIfClosed() {
        if (closed) throw new IllegalStateException("IdSMTPMin already closed.");
    }
}
